webextension/background.js

```diff
@@ -0,0 +1,1006 @@
+  browser.webNavigation.onCompleted.addListener(onNavigationCompleted);
+  browser.study.onEndStudy.addListener(deactivate);
+  active = true;
+}
+
+function deactivate() {
+  if (!active) {
+    return;
+  }
+  active = false;
+  VisitTimeTracker.stop();
+  hidePageActionOnEveryTab();
+  gCurrentlyPromptingTab = undefined;
+  browser.tabs.onActivated.removeListener(onTabChanged);
+  browser.webNavigation.onCommitted.removeListener(onNavigationCommitted);
+  browser.webNavigation.onCompleted.removeListener(onNavigationCompleted);
+  browser.management.uninstallSelf();
+}
+
+Config.load().then(activate).catch(loadingError => {
+  if (Config.testingMode) {
+    console.info("Failed to start addon; uninstalling:", loadingError);
+  }
+  browser.management.uninstallSelf();
+});
+
+function hidePageActionOnEveryTab() {
+  browser.tabs.query({}).then(tabs => {
+    for (const {id} of tabs) {
+      browser.pageAction.hide(id);
+    }
+  });
+}
+
+async function handleButtonClick(command, tabState) {
+  switch (tabState.slide) {
+    case "initialPrompt": {
+      const siteWorks = command === "yes";
+      tabState.maybeSendTelemetry({satisfiedSitePrompt: yesOrNo(siteWorks)});
+      if (siteWorks) {
+        tabState.slide = "thankYou";
+        tabState.markAsVerified();
+      } else {
+        tabState.slide = "feedbackForm";
+      }
+      break;
+    }
+    case "feedbackForm": {
+      if (command === "submit") {
+        tabState.submitReport();
+        tabState.slide = "thankYouFeedback";
+      } else if (command === "showFeedbackDetails") {
+        tabState.slide = "feedbackDetails";
+        tabState.maybeUpdatePageAction(["type", "description"]);
+      } else if (command === "back") {
+        tabState.slide = "initialPrompt";
+      } else if (command === "cancel") {
+        closePageAction();
+        tabState.reset();
+        tabState.maybeSendTelemetry({shareFeedBack: "userCancelled"});
+      }
+      tabState.markAsVerified();
+      break;
+    }
+    case "feedbackDetails": {
+      if (command === "submit") {
+        tabState.submitReport();
+        tabState.slide = "thankYouFeedback";
+        tabState.markAsVerified();
+      } else if (command === "back") {
+        tabState.slide = "feedbackForm";
+      } else if (command === "cancel") {
+        closePageAction();
+        tabState.reset();
+        tabState.markAsVerified();
+        tabState.maybeSendTelemetry({shareFeedBack: "userCancelled"});
+      }
+      break;
+    }
+    case "thankYou": {
+      closePageAction();
+      break;
+    }
+    case "thankYouFeedback": {
+      closePageAction();
+      break;
+    }
+  }
+}
+
+async function updatePageActionIcon(tabId) {
+  const active = (gCurrentlyPromptingTab || {}).id === tabId;
+  const path = active ? "icons/broken_page_active.svg"
+                      : "icons/broken_page.svg";
+  await browser.pageAction.setIcon({tabId, path});
+}
+
+function getActiveTab() {
```

Comment:
You're already listening to `tabs.onActivated` so you should be able to keep track of this and then not need an async call here (though you'd actually need to keep track of the active tab per-window and also listen for windows focus changes)

Reply:
I'll definitely be considering changes like this once the experiment is over, but for now I'd rather just do as async call than manage more state than necessary.